File: variants/zig/src/rca/thread.zig
///////////////////////////////////////////////////////////////////////////////
// Regulated Cell Architecture (RCA) - Zig Variant
// Thread
///////////////////////////////////////////////////////////////////////////////

const data = @import("data.zig");
const cell_mod = @import("cell.zig");

const DataPlane = data.DataPlane;
const ActivityInfo = data.ActivityInfo;
const Cell = cell_mod.Cell;
const CellData = cell_mod.CellData;
const CELLS = cell_mod.CELLS;
const cellDefaults = cell_mod.cellDefaults;

// *****************************************************************************
// (1) Threads
// *****************************************************************************

// Status: MUTABLE
pub const THREADS: usize = 1;

// Status: MUTABLE
pub const EXECUTION_THRESHOLD: f64 = 1.0; // Units in ms

// Status: MUTABLE
pub const Effect = struct {
    activity: ActivityInfo,
    handoff: *const CellData,
    finished: bool,
};

fn taskName(task: cell_mod.Task) []const u8 {
    return switch (task) {
        .default => "Default",
        .double_value => "DoubleValue",
    };
}

// Status: MUTABLE
pub const ProgramThread = struct {
    counter: usize = 0,
    tasks: [CELLS]Cell = cellDefaults(),
    handoff: CellData = .none,

    pub fn buildTasks(tasks: ?[CELLS]Cell, ho: ?CellData) ProgramThread {
        return .{
            .counter = 0,
            .tasks = tasks orelse cellDefaults(),
            .handoff = ho orelse .none,
        };
    }

    pub fn isFinished(self: *const ProgramThread) bool {
        return self.counter >= CELLS;
    }

    pub fn step(self: *ProgramThread, ctx: *const DataPlane) Effect {
        const activity = ActivityInfo{
            .description = taskName(self.tasks[self.counter].task),
        };

        // Handoff transfer: take current, pass to cell, store result back
        const handoff_transfer = self.handoff;
        self.handoff = .none;

        self.handoff = self.tasks[self.counter].execute(ctx, handoff_transfer);
        self.counter += 1;

        const finished = self.isFinished();

        return .{
            .activity = activity,
            .handoff = &self.handoff,
            .finished = finished,
        };
    }
};

File: variants/c/rca/thread.c
/*******************************************************************************
 * Regulated Cell Architecture (RCA) - C Variant
 * Thread Implementation
 ******************************************************************************/

#include "thread.h"
#include <stdio.h>
#include <string.h>

static const char *task_name(Task t) {
    switch (t) {
        case TASK_DEFAULT:      return "Default";
        case TASK_DOUBLE_VALUE: return "DoubleValue";
    }
    return "Unknown";
}

void program_thread_build(ProgramThread *pt, const Cell *tasks, const CellData *handoff) {
    pt->counter = 0;

    if (tasks) {
        memcpy(pt->tasks, tasks, sizeof(Cell) * CELLS);
    } else {
        cell_defaults(pt->tasks);
    }

    if (handoff) {
        pt->handoff = *handoff;
    } else {
        cell_data_default(&pt->handoff);
    }
}

bool program_thread_is_finished(const ProgramThread *pt) {
    return pt->counter >= CELLS;
}

int program_thread_step(ProgramThread *pt, const DataPlane *ctx, Effect *effect) {
    /* Build activity description */
    snprintf(effect->activity.description, sizeof(effect->activity.description),
             "%s", task_name(pt->tasks[pt->counter].task));

    /* Handoff transfer: take current handoff, pass to cell, store result back */
    CellData handoff_transfer = pt->handoff;
    cell_data_default(&pt->handoff);

    int rc = cell_execute(&pt->tasks[pt->counter], ctx, handoff_transfer, &pt->handoff);
    if (rc != 0) {
        return rc;
    }

    pt->counter++;

    effect->finished = program_thread_is_finished(pt);
    effect->handoff  = &pt->handoff;

    return 0;
}
